enclave-runtime/src/lib.rs

```diff
@@ -597,119 +604,186 @@ where
 	let tx_pool_guard = pool_mutex.lock().unwrap();
 	let tx_pool = tx_pool_guard.deref();
 
-	// Handle trusted getters
-	execute_trusted_getters(tx_pool, &shards)?;
+	// execute pending calls from operation pool and create block
+	for shard in shards.into_iter() {
+		let shard_exec_time = match remaining_time(ends_at)
+			.map(|r| r.checked_div(remaining_shards))
+			.flatten()
+		{
+			Some(t) => t,
+			None => {
+				info!("[Enclave] Could not execute trusted operations for all shards. Remaining shards: {}.", remaining_shards);
+				break
+			},
+		};
+
+		match exec_tops::<PB, SB, _, _>(
+			ocall_api,
+			tx_pool,
+			&latest_onchain_header,
+			shard,
+			shard_exec_time,
+		) {
+			Ok((confirm_calls, sb)) => {
+				calls.extend(confirm_calls);
+				if let Some(sb) = sb {
+					signed_blocks.push(sb);
+				}
+			},
+			Err(e) => error!("Error in top execution for shard {:?}: {:?}", shard, e),
+		}
+
+		remaining_shards -= 1;
+	}
+	Ok((calls, signed_blocks))
+}
+
+/// Execute pending trusted operations for the `shard` until the `max_exec_duration` is reached.
+///
+/// The first half of the `max_exec_duration` is dedicated to the trusted getters, the second half
+/// (plus leftover time from the getters) to the trusted calls.
+///
+/// Todo: The getters should be handled individually: #400
+fn exec_tops<PB, SB, O, P>(
+	ocall_api: &O,
+	top_pool: &P,
+	latest_onchain_header: &PB::Header,
+	shard: ShardIdentifier,
+	max_exec_duration: Duration,
+) -> Result<(Vec<OpaqueCall>, Option<SB>)>
+where
+	PB: BlockT<Hash = H256>,
+	SB: SignedBlockT<Public = sp_core::ed25519::Public, Signature = MultiSignature>,
+	SB::Block: SidechainBlockT<ShardIdentifier = H256, Public = sp_core::ed25519::Public>,
+	O: EnclaveOnChainOCallApi,
+	P: TrustedOperationPool<Hash = H256> + 'static,
+{
+	// first half of the slot is dedicated to getters.
+	let ends_at = duration_now() + max_exec_duration;
+	let remaining_getter_time = max_exec_duration
+		.checked_div(2)
+		.expect("checked_div yields some if rhs != 0; qed");
```

Comment:
is there no 'const' version of `div`, where already at compile-time the divisor is guaranteed to be != 0?

Reply:
Haha, you are so damn right. I was completely focused on removing the unwrap. 😄

Reply:
I'm not sure if it's possible though in Rust - but maybe you know more?

Reply:
Aha, I think I misunderstood you. I simply replaced the `checked_div` with `/`, which would panic if the rhs is 0.

Reply:
That is fine too I think 👍 I was thinking about a `checked_div` variant, where the divisor is a compile-time constant, so the compiler can already ensure there is never a division by 0. But in my very brief research on this, I couldn't find anything.

Reply:
Oh, I see. I am not entirely sure.

But inside the `/` operator, the panic is triggered in an if statement. I believe that the compiler is, in general, smart enough to skip if-branches that are never taken. I don't remember where I read that. I only found this question that goes somewhat in this direction: https://users.rust-lang.org/t/has-rust-compiler-made-any-optimization-for-conditional-operations/16436